### PR TITLE
Support `@serdeEnumProxy` (`@serdeProxyCast`) UDA

### PR DESCRIPTION
**Pull request by WebFreak001**
depends on https://github.com/libmir/mir-algorithm/pull/408

### CODE REVIEW

**Comment by WebFreak001** [comment]
marking as draft to explore different PR right now

**Comment by WebFreak001** [comment]
not sure if integration test failures are related, the code changes here are entirely compile-time opt-in by annotating enums with an UDA.

Some DMD tests need restarting because of download failure.

File: source/mir/deser/package.d [hunk 1]
```diff
@@ -802,16 +802,33 @@ template deserializeValue(string[] symbolTable, TableKind tableKind, bool annota
                 if (auto exception = impl(data, temporal, runtimeSymbolTable, compiletimeIndex, annotations_))
                     return exception;
 
-                static if (__traits(compiles, ()@safe{return to!T(move(temporal));}))
-                    value = to!T(move(temporal));
+                static if (hasUDA!(T, serdeProxyCast))
+                {
+                    static if (__traits(compiles, ()@safe{return cast(T)move(temporal);}))
```
Comment (9il) [inline source/mir/deser/package.d]: `cast(T)move ( ...` <- `move` is useless when cast. Please remove

File: source/mir/ser/package.d [hunk 2]
```diff
@@ -107,7 +107,10 @@ void serializeValue(S, V)(scope ref S serializer, scope const V value)
 {
     static if (hasUDA!(V, serdeProxy))
     {
-        serializer.serializeWithProxy!(serdeGetProxy!V)(value);
+        static if (hasUDA!(V, serdeProxyCast))
+            serializeValue(serializer, cast(serdeGetProxy!V)value);
```
Comment (9il) [inline source/mir/ser/package.d]: The cast may return by value. `serializeValue` may want a reference.

File: source/mir/ser/package.d [hunk 3]
```diff
@@ -107,7 +107,10 @@ void serializeValue(S, V)(scope ref S serializer, scope const V value)
...
+        else
+            serializer.serializeWithProxy!(serdeGetProxy!V)(value);
```
Comment (9il) [inline source/mir/ser/package.d]: `serializeWithProxy` has to be updated instead with all its calls and test for them. Please add tests for those cases too.

**Pull request merged**